servers/cu/src/domain/client/ao-process.test.js

```diff
@@ -1396,5 +1396,107 @@ describe('ao-process', () => {
     })
   })
 
-  describe.todo('saveCheckpointWith')
+  describe('saveCheckpointWith', () => {
+    const logger = createLogger('saveCheckpointWith')
+    const depsAll = {
+      logger
+    }
+
+    test('should not save checkpoint if checkpointing is disabled', async () => {
+      const saveCheckpoint = saveCheckpointWith({
+        ...depsAll,
+        address: () => {
+          assert.fail('saving checkpoint process should not have been initiated')
+        },
+        DISABLE_PROCESS_CHECKPOINT_CREATION: true
+      })
+      await saveCheckpoint({})
+    })
+
+    test('should not save checkpoint if process was recently checkpointed', async () => {
+      const saveCheckpoint = saveCheckpointWith({
+        ...depsAll,
+        address: () => {
+          assert.fail('saving checkpoint process should not have been initiated')
+        },
+        DISABLE_PROCESS_CHECKPOINT_CREATION: false,
+        recentCheckpoints: new Map([['process-123', true]])
+      })
+      await saveCheckpoint({ processId: 'process-123' })
+    })
+
+    test('should not save a checkpoint if there is already a checkpoint for the evaluation on Arweave', async () => {
+      const saveCheckpoint = saveCheckpointWith({
+        ...depsAll,
+        address: () => {
+          return Promise.resolve()
+        },
+        queryGateway: () => Promise.resolve({ data: { transactions: { edges: [{ node: { id: 'tx-123', tags: [] } }] } } }),
+        queryCheckpointGateway: () => Promise.resolve({ data: { transactions: { edges: [{ node: { id: 'tx-123', tags: [] } }] } } }),
+        readProcessMemoryFile: () => {
+          assert.fail('should not attempt to read the process memory file')
+        },
+        writeCheckpointRecord: () => Promise.resolve(),
+        DISABLE_PROCESS_CHECKPOINT_CREATION: false
+      })
+      const result = await saveCheckpoint({ File: Promise.resolve('file') })
+      assert.equal(result.id, 'tx-123', 'should return the id of the checkpoint')
+    })
+
+    describe('should fail if the memory is not provided directly or as a field path', () => {
+      const deps = {
+        ...depsAll,
+        address: () => {
+          return Promise.resolve()
+        },
+        queryGateway: () => Promise.resolve(),
+        queryCheckpointGateway: () => Promise.resolve(),
+        DISABLE_PROCESS_CHECKPOINT_CREATION: false
+      }
+      test('should fail if memory and file are both undefined', async () => {
+        const saveCheckpoint = saveCheckpointWith({
+          ...deps
+        })
+        try {
+          await saveCheckpoint({})
```

Comment:
this should have something like .then(() => assert.fail('should not resolve if no Memory or File is provided')) or something like that.